### PR TITLE
Plugin plots to viewers

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request migrates all plugin plots (plot options stretch histogram, imviz's line profile, imviz's aperture photometry radial profile) to use a glue-viewer object, with their own independent app instances, instead of a bqplot object.  This then allows us to use our nested toolbar on these plots (note that the data are not linked to the main data-collection, so using subset tools probably doesn't make sense).  This is left somewhat flexible so that we can share the app instance when it makes sense - possibly for the compass plugin in the future.

For the stretch histogram, this _removes_ the need for downsampling, but also loses the logic for ensuring the vmin/vmax markers are visible (since the limit state is handled by the viewer directly and the user can override - whereas previously the limits were fixed by code in the plugin).

Note that the histogram implementation includes some hacky workarounds due to bugs upstream.  I'll work on reporting those and we can create tech debt tickets to remove the workarounds.  The home button on the histogram is also known to be broken - we can either live with that until there is an upstream fix, remove the home button for the histogram in this PR, or block merging this PR until there is an upstream fix.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by pllim** [comment]
Is this the one that segfaults with fast-histogram? Should I hold off on reviewing?

**Comment by kecnry** [comment]
> Is this the one that segfaults with fast-histogram? Should I hold off on reviewing?

It is, but I am quite sure that is an upstream bug (that will need to be addressed before merging).  If it doesn't segfault for you, then I think its ready for reviewing!  And if it does - the code can still be reviewed.

We also need to make decisions on the other "bugs" in the histogram viewer and whether those block merge or not, given that we want to build on top of this very soon.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:395]: What is the point of emitting snackbar message if you are going to throw the exception here anyway?

**Review by pllim — COMMENTED** [review]
I also got the segfault like CI and Ricky, so I cannot run the code, but I left some general comments.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:395]: Woops, this was just for debugging and I forgot to remove it. Thanks for noticing!

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:69]: Nothing yet, I've asked if it is intentionally excluded and if it isn't I'll open a PR there.

**Comment by rosteen** [comment]
This is pretty cool, I like it more than I thought I would. That said, I've encountered two issues. The first is that in the histogram viewer, the stretch curve does not seem to update based on panning and zooming correctly:


https://github.com/spacetelescope/jdaviz/assets/39831871/15e3642b-6b3b-46fd-a18d-7a0dd2d0892f

And the second is that nothing plots for me in the line profile plugin:


https://github.com/spacetelescope/jdaviz/assets/39831871/af4cf58e-3275-4c44-b7ed-6fa245f3beeb

I took a cursory read through the code changes and nothing egregious stood out to me, but I'll try to take a closer look before actually approving. I figured I'd bring these bugs up now though.

**Comment by kecnry** [comment]
> The first is that in the histogram viewer, the stretch curve does not seem to update based on panning and zooming correctly

This is by-design (but can be debated if it's what we want).  The stretch function is not in the same units as the y-axis of the histogram, and so is essentially tied to a separate, normalized, y-axis.  I thought about only allowing horizontal pan/zooming to avoid any confusion here, but I think it is useful to be able to zoom vertically on the tail of the histogram as well.

> And the second is that nothing plots for me in the line profile plugin

hmmm, I can't reproduce that 🤔  Does something show if you press the home button?

**Comment by rosteen** [comment]
> Does something show if you press the home button?

Nope, nothing.

👍 on the histogram, maybe demo that offline at the review today and get feedback? I don't think it's super important but could potentially be confusing.

**Comment by kecnry** [comment]
Bizarre, can you dig into the object itself and see if the arrays are populated as expected?

```
lp = imviz.plugins['Imviz Line Profiles (XY)']
plot = lp._obj.plot_across_x

data = plot.app.data_collection[0]
print(data.get_component('x').data, data.get_component('y').data)

line_layer = plot.layers['line']
print(line_layer.layer.data['x'], line_layer.layer.data['y'])
```

**Comment by kecnry** [comment]
> on the histogram, maybe demo that offline at the review today and get feedback? I don't think it's super important but could potentially be confusing.

Some feedback from review: this probably makes sense as-is for now, but we might need to reconsider when adding support for custom/interactive splines.  The orange vertical line for vmin/vmax also looks like its a bug, so a different color might help (that we can do when making them interactive as well).

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:69]: https://github.com/glue-viz/glue-jupyter/pull/402

File: jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: Tsk tsk... icon needs dark mode version. 😉 

Also Y-axis for Line Profile plot probably needs scientific notation.

Screenshot 2023-10-10 144451

File: jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:128]: This block and similar block below that were deleted need to be restored but in your new way, so the Line Profile plots respect image zoom limits.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1]: Curve of Growth and Radial Profile need a gray line connecting the dots.

Screenshot 2023-10-10 145741

Radial Profile (Raw) is not showing anything. But no error either. Not sure what is happening. If you plot the fit, you are supposed to see a bunch of points around the line. If you do not plot the fit, you are supposed to see the points as scatter plot (no line connecting them). But I do not see the points at all now. If I do not plot the fit, this plot below is just blank.

Screenshot 2023-10-10 150012

Like Line Profile, the Y-axis label values need scientific notation.

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:7]:
```suggestion
- Plots in plugins now include basic zoom/pan tools for Plot Options,
  Imviz Line Profiles, and Imviz Simple Aperture Photometry. [#2498]
```

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:1]: Like Line Profile, the histogram here is also not responding to me zooming in/out of the image. I thought it was supposed to?

Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1]: Ah, never mind. For this one, I forgot there is a toggle. Ignore this comment.

(In Line Profile, it is supposed to be automatic.)

Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:1]: For Mosviz, you might want the histogram to always follow zoom limits. Otherwise, in `MosvizNIRISSExample.ipynb`, the histogram won't update when you switch row, which can be confusing. If this is a separate ticket, lemme know.

**Review by pllim — COMMENTED** [review]
Overall, I like this but the plots in Aperture Photometry and Line Profile have to be fixed first.

File: jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: Personally, I expected it to initially follow the zoom, but then if I additionally pan/zoom it, then it is on me. But then again, pressing "l" around the same pixel would ideally give me the default behavior back, so no big deal.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1]: was this supported before?  If not, I'd say out of scope here.

File: jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: So after you press "l" changes to the image viewer zoom would have a one-way sync back to here, and once you use the zoom tool here, that sync would stop until you press "l" again.  I worry without visual indication, that that will be quite confusing.  We could maybe hook into the matched zoom tools that we already use and have (optional) 2-way sync.

If we're not ready to make a decision yet, I can re-implement the former logic and _remove_ the zoom tool from line profiles for now.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1]: > Radial Profile (Raw) is not showing anything. But no error either. Not sure what is happening

Interesting, I see them on my end.  I wonder if the point size might just be too small.  Can you try changing `self.plot.update_style('profile', color='gray', size=1)` to a larger size and see if that works?

I'll work on the other changes to styling to bring them more in line with the previous version.  Thanks for pointing those out.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:1]: Okay, I opened https://github.com/spacetelescope/jdaviz/issues/2501

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1]: I think I've made all your requested styling changes (including dark mode support).  Please let me know if I missed anything.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:1]: Dark mode looks fine now and I see the line connecting dots, thanks!

Have to do this to make the raw profile show up:

```diff
--- a/jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
+++ b/jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
@@ -356,7 +356,7 @@ class SimpleAperturePhotometry(PluginTemplateMixin, DatasetSelectMixin, TableMix
                         raw=True)

                     self.plot._update_data('profile', x=x_data, y=y_data, reset_lims=True)
-                    self.plot.update_style('profile', line_visible=False, color='gray', size=1)
+                    self.plot.update_style('profile', line_visible=False, color='gray', size=10)
```

Screenshot 2023-10-10 175602

File: jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: I think 2-way sync is overkill. It is fine if you want to remove the zoom tool. But we definitely want to keep at least the initial line profile axes to match main viewer zoom when "l" is pressed. Thanks!

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: ok, as long as it doesn't then sync (even one way) after, that shouldn't cause conflicts.  I'll work on re-implementing that.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/line_profile_xy/line_profile_xy.py:1]: I agree. Just the initial axes matching zoom. It should no longer sync to anything after that.

**Comment by rosteen** [comment]
> @rosteen - can you check to see if increasing the size on the line profile lines makes them visible on your machine?

Doesn't seem like it, unfortunately.

File: CHANGES.rst
Comment (kecnry) [reply CHANGES.rst:7]: is this just because of concern it would otherwise suggest the plugin?  (I don't mind being more explicit, but #2465 is also changing the name of the aperture photometry plugin and so we'd have to remember there to update this as well).

**Comment by pllim** [comment]
I don't see Ricky's problems but I also rebased this PR on top of latest main before I installed it.

I see that the histogram stretch function does not update when I have it shown and then press the Home button but I think you said that is a known issue?

**Comment by kecnry** [comment]
> I see that the histogram stretch function does not update when I have it shown and then press the Home button but I think you said that is a known issue?

Yes, the home button (or calling reset limits from the API) for histograms breaks.  As mentioned in the description, we can choose to live with that for now as a known bug until it is fixed upstream, remove the home button for that plot here, or block merging this PR until we have the fix.

**Comment by kecnry** [comment]
line profile not showing should now be fixed by bumping glue-jupyter to 0.19 (which added support for lines in scatter viewers.... so obvious in hindsight).  Anyone testing, make sure you have updated glue-jupyter!

**Comment by pllim** [comment]
Re: https://github.com/spacetelescope/jdaviz/pull/2498#issuecomment-1757747042

Removing or disabling HOME for histogram might be good, to prevent people from sending in help call about something we know is broken. Otherwise, we can also just live with it. I don't think it should block merging.

File: jdaviz/core/template_mixin.py
Comment (pllim) [inline jdaviz/core/template_mixin.py:3264]: I vaguely remember @astrofrog said we should do something like this? I see it used in Imviz helper.

```suggestion
            with self.app.data_collection.delay_link_manager_update():
                dc.add_link(links)
```

**Review by pllim — COMMENTED** [review]
Works as expected now but the axes don't look quite right even in pop-out.

Screenshot 2023-10-11 151117

1. The Y-axis label is on top of the values. I think we need to move the axes back a little bit to the right. This affects both aperture photometry and line profile plugins.
2. The X-axis label value get cut off a little on the right edge, so might need more padding. I only see this for Line Profile plugin.

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:3264]: does this do anything if there is only a _single_ call to `add_link`?

Comment (pllim) [reply jdaviz/core/template_mixin.py:3264]: Hmm maybe not...? https://jdaviz.readthedocs.io/en/latest/dev/links.html#speeding-up-adding-links-to-the-data-collection

**Review by pllim — APPROVED** [review]
LGTM now.

I did see this weird thing where I thought I saw the scatter plot points flashed real quick but then they disappeared from the plot, but then if I pop the same plot out, I see them in the pop-out only. But I cannot reproduce it, so maybe we can chuck it under Heisenbug.

Screenshot 2023-10-11 173213

**Review by rosteen — APPROVED** [review]
Works for me now, code looks fine. Let's get this in so we can iterate on it.

**Comment by kecnry** [comment]
Ok, please hold off on merging for a bit - I'll merge aperture photometry first since that should be an easier order for the rebase 🤞

**Pull request merged**
